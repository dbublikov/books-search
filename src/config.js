import axios from 'axios';

const API_KEY = 'AIzaSyB0Gf_r2La1jo5bMrpUzgVaQqHkxqS05-A';

const baseUrl = 'https://www.googleapis.com/books/v1/volumes';

const getBooks = `${baseUrl}?q=subject:art&orderBy=newest&maxResults=12&key=${API_KEY}`;

export const initLoad = async () => {
  const res = await axios.get(getBooks);
  // console.log(res)
  if (res.isAxiosError) {
    throw new Error('Can\'t fetch');
  }
  return res;
};

export const loadDetails = async (id) => {
  const res = await axios.get(`${baseUrl}/${id}`);
  if (res.isAxiosError) {
    throw new Error('Can\'t fetch');
  }
  return res;
};
