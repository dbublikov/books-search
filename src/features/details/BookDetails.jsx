import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { Info } from './Info';
import { Loader } from '../../components/Loader';
import { clearDetails, getDetails, loadBookById, selectDetails } from './details-slice';

export const BookDetails = ({ id }) => {
  const { status, error, currentBook } = useSelector(selectDetails);
  const dispatch = useDispatch();

  useEffect(() => {
    if (localStorage.hasOwnProperty(id)) {
      // console.log('check localStorage!!!')
      dispatch(getDetails(id));
    } else {
      dispatch(loadBookById(id));
    }

    return () => {
      dispatch(clearDetails());
    };
  }, [id, dispatch]);

  return (
    <>
      {status === 'loading' && <Loader pos={true} />}
      {error && <Loader>Can't fetch the book info =(</Loader>}
      {currentBook && <Info id={id} {...currentBook.volumeInfo} />}
    </>
  );
};
