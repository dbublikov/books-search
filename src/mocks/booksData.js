export const booksData = {
  items: [
    {
      id: 'osZVEAAAQBAJ',
      volumeInfo: {
        title: 'Start Painting Now',
        subtitle: 'Discover Your Artistic Potential',
        authors: [
          'Emily Powell',
          'Sarah Moore',
        ],
      },
    },
    {
      id: 'd_E3EAAAQBAJ',
      volumeInfo: {
        title: 'What the Ermine Saw',
        subtitle: 'The Extraordinary Journey of Leonardo da Vinci\'s Most Mysterious Portrait',
        authors: [
          'Eden Collinsworth',
        ],
      },
    },
    {
      id: 'gwJLEAAAQBAJ',
      volumeInfo: {
        title: 'What Artists Wear',
        authors: [
          'Charlie Porter',
        ],
      },
    },
  ],
  totalItems: 6,
};

export const extraBooksData = {
  items: [
    {
      id: '6DBNEAAAQBAJ',
      volumeInfo: {
        title: 'Looking for Theophrastus',
        subtitle: 'Travels in Search of a Lost Philosopher',
        authors: [
          'Laura Beatty',
        ],
      },
    },
    {
      id: 'tYkqEAAAQBAJ',
      volumeInfo: {
        title: 'Portrait of a Thief',
        authors: [
          'Grace D. Li',
        ],
      },
    },
    {
      id: 'xsNEEAAAQBAJ',
      volumeInfo: {
        title: 'The Art of Cloth in Mughal India',
        authors: [
          'Sylvia Houghteling',
        ],
      },
    },
  ],
};

