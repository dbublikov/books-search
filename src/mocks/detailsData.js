export const detailsData = {
  id: '2upIEAAAQBAJ',
  volumeInfo: {
    authors: ['Hugh Eakin'],
    title: 'Picasso\'s War',
    subtitle: 'How Modern Art Came to America',
    categories: ['Art / History / Modern (late 19th Century to 1945)'],
    description: '<b><i>NEW YORK TIMES</i> EDITORS’ CHOICE • A riveting story of how dueling ambitions and the power of prodigy made America the cultural center of the world—and Picasso the most famous artist alive—in the shadow of World War II<br><br>“Fascinating, eloquent, wonderfully lucid, <i>Picasso’s War</i> will change whatever we thought we knew about modern art and its complicated reception on this side of the Atlantic.”—Francine Prose, author of <i>The Vixen</i></b>',
    pageCount: 480,
    publisher: 'Crown',
    publishedDate: '2022-07-12',
  },
};


