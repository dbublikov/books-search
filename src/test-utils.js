import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';
import { controlsReducer } from './features/controls/controls-slice';
import { detailsReducer } from './features/details/details-slice';
import { booksReducer } from './features/books/books-slice';

const reducer = {
  books: booksReducer,
  controls: controlsReducer,
  details: detailsReducer,
};

export function getStoreWithState(preloadedState) {
  return configureStore({ reducer, preloadedState });
}

export function renderWithContext(element, state) {
  const store = getStoreWithState(state);
  const utils = render(
    <Provider store={store}>
      <Router>{element}</Router>
    </Provider>
  );
  return { store, ...utils };
}

export const getPreloadedState = () => {
  const state = {
    books: {
      status: 'idle',
      error: null,
      totalItems: null,
      list: [],
      pagination: {
        status: 'idle',
        error: null,
        currIndex: 2,
        pageTerm: 'placeholder',
        step: 2,
      },
    },
  };
  return state;
};
