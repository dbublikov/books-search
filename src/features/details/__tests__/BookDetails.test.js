import { screen, waitFor } from '@testing-library/react';
import { renderWithContext } from '../../../test-utils';

import { BookDetails } from '../BookDetails';
import { detailsData } from '../../../mocks/detailsData';
import * as api from '../../../config';

const getBookSpyDetails = jest.spyOn(api, 'loadDetails');

describe('BookDetails', () => {
  it('should render book details', async () => {
    getBookSpyDetails.mockResolvedValue({ data: detailsData });
    const view = renderWithContext(<BookDetails />);

    expect(screen.getByTestId('spinner')).toBeVisible();
    await waitFor(() => expect(getBookSpyDetails).toHaveBeenCalledTimes(1));
    expect(await screen.findByText(/Picasso's War/i)).toBeInTheDocument();
    expect(await screen.findByText(/Hugh Eakin/i)).toBeInTheDocument();
    expect(view).toMatchSnapshot();
  });

  it('should render error message', async () => {
    getBookSpyDetails.mockRejectedValue(new Error('API Error'));
    renderWithContext(<BookDetails />);

    await waitFor(() => expect(getBookSpyDetails).toHaveBeenCalledTimes(1));
    expect(await screen.findByText(/Can't fetch the book info/i)).toBeInTheDocument();
  });
});





