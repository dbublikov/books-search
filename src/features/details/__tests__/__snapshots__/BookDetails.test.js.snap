// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetails should render book details 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="sc-eDvSVe fnPPbA"
      >
        <div
          class="sc-jSUZER kJIKHL"
        >
          <img
            alt="Picasso's War"
            class="sc-gKPRtg glIhiL"
            src="https://books.google.com/books/publisher/content/images/frontcover/undefined?fife=w480-h960"
          />
        </div>
        <div>
          <h3
            class="sc-iBYQkv jvPpqk"
          >
            Picasso's War
          </h3>
          <h5
            class="sc-ftTHYK dlrzcd"
          >
            Art / History / Modern (late 19th Century to 1945)
          </h5>
          <h5
            class="sc-pyfCe kSUTtz"
          >
            Hugh Eakin
          </h5>
          <h5
            class="sc-jrcTuL eLiLQF"
          >
            Crown
            , 
            2022-07-12
            , 
             pages
          </h5>
          <div
            class="sc-kDvujY cDehyk"
          >
            <b>
              <i>
                NEW YORK TIMES
              </i>
               EDITORS’ CHOICE • A riveting story of how dueling ambitions and the power of prodigy made America the cultural center of the world—and Picasso the most famous artist alive—in the shadow of World War II
              <br />
              <br />
              “Fascinating, eloquent, wonderfully lucid, 
              <i>
                Picasso’s War
              </i>
               will change whatever we thought we knew about modern art and its complicated reception on this side of the Atlantic.”—Francine Prose, author of 
              <i>
                The Vixen
              </i>
            </b>
          </div>
          <div
            class="sc-ipEyDJ kLdPaW"
          >
            <a
              class="sc-csuSiG BAxNQ"
              rel="noreferrer"
              target="_blank"
            >
              <button
                class="sc-bcXHqe QdSeH"
              >
                Preview
              </button>
            </a>
            <a
              class="sc-csuSiG BAxNQ"
              rel="noreferrer"
              target="_blank"
            >
              <button
                class="sc-bcXHqe QdSeH"
              >
                Info
              </button>
            </a>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="sc-eDvSVe fnPPbA"
    >
      <div
        class="sc-jSUZER kJIKHL"
      >
        <img
          alt="Picasso's War"
          class="sc-gKPRtg glIhiL"
          src="https://books.google.com/books/publisher/content/images/frontcover/undefined?fife=w480-h960"
        />
      </div>
      <div>
        <h3
          class="sc-iBYQkv jvPpqk"
        >
          Picasso's War
        </h3>
        <h5
          class="sc-ftTHYK dlrzcd"
        >
          Art / History / Modern (late 19th Century to 1945)
        </h5>
        <h5
          class="sc-pyfCe kSUTtz"
        >
          Hugh Eakin
        </h5>
        <h5
          class="sc-jrcTuL eLiLQF"
        >
          Crown
          , 
          2022-07-12
          , 
           pages
        </h5>
        <div
          class="sc-kDvujY cDehyk"
        >
          <b>
            <i>
              NEW YORK TIMES
            </i>
             EDITORS’ CHOICE • A riveting story of how dueling ambitions and the power of prodigy made America the cultural center of the world—and Picasso the most famous artist alive—in the shadow of World War II
            <br />
            <br />
            “Fascinating, eloquent, wonderfully lucid, 
            <i>
              Picasso’s War
            </i>
             will change whatever we thought we knew about modern art and its complicated reception on this side of the Atlantic.”—Francine Prose, author of 
            <i>
              The Vixen
            </i>
          </b>
        </div>
        <div
          class="sc-ipEyDJ kLdPaW"
        >
          <a
            class="sc-csuSiG BAxNQ"
            rel="noreferrer"
            target="_blank"
          >
            <button
              class="sc-bcXHqe QdSeH"
            >
              Preview
            </button>
          </a>
          <a
            class="sc-csuSiG BAxNQ"
            rel="noreferrer"
            target="_blank"
          >
            <button
              class="sc-bcXHqe QdSeH"
            >
              Info
            </button>
          </a>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
